### PR TITLE
Length of modules

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
We compute the length of a finite module over a local ring and an associated decomposition series.

### CODE REVIEW

File: test/Modules/local_rings.jl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  N, _ = quo(F, (ideal(L, [x,y,z])*F)[1])
+  @test length(N) == 0
+
+  @test_throws "module does not have finite length" length(F)
+
+  A, _ = quo(R, ideal(R, [x^2, y, z]))
+  W, _ = localization(A, complement_of_ideal(ideal(R, [x,y,z])))
+  F = FreeMod(W, 2)
+  @test_broken length(F) == 4 # See #1806
```
Comment (HechtiDerLachs) [inline test/Modules/local_rings.jl]: This should indeed be fixed by that branch.

File: src/Modules/local_rings.jl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  N, _ = quo(F, (R(P)*F)[1])
+  k = 0
+  MM = M
+  p = identity_map(M)
+  NtoM, interp = hom(N, MM)
+  decomp = Vector{elem_type(M)}()
+  while !iszero(NtoM)
+    g = gens(NtoM)
+    i = findfirst(x->!(iszero(x)), g)
+    Msub, inc = sub(MM, [interp(g[i])(N[1])])
```
Comment (afkafkafk13) [inline src/Modules/local_rings.jl:21]: I know that this remark has nothing to do with this pull-request, but 'interp' is a non-intuitive name....

File: src/Modules/local_rings.jl [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    return 0
+  end
+  R = base_ring(M)
+  P = prime_ideal(inverted_set(R))
+  F = FreeMod(R, 1)
+  N, _ = quo(F, (R(P)*F)[1])
+  k = 0
+  MM = M
+  p = identity_map(M)
+  NtoM, interp = hom(N, MM)
```
Comment (afkafkafk13) [inline src/Modules/local_rings.jl:21]: interp is a non-intuitive name for this map, but I do not have a better name off-hand either.

File: src/Modules/local_rings.jl [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    p = compose(p, pp)
+    NtoM, interp = hom(N, MM)
+    k = k + 1
+  end
+  iszero(MM) || error("module does not have finite length")
+  set_attribute!(M, :decomposition_series, decomp)
+  return k
+end
+
+function decomposition_series(
```
Comment (afkafkafk13) [inline src/Modules/local_rings.jl]: Do you want a 'composition series' ?  (question on terminology)

File: src/Modules/local_rings.jl [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    M::ModuleFP{RingElemType}
+  ) where {RingElemType<:AbsLocalizedRingElem{<:Any, <:Any, <:MPolyComplementOfPrimeIdeal}}
+  if !has_attribute(M, :decomposition_series)
+    length(M)
```
Comment (afkafkafk13) [inline src/Modules/local_rings.jl]: Using this sideeffect of length does not seem structurally honest.... Should length not be relying on composition_series in the case of localizations at complements of prime ideals?
(by the way: in what file is the case of localization at MPolyComplementOfKPointIdeal handled? That one should not pass through a composition series, but use combinatorics on the leading module -- and hence not have this sideeffect.)

File: src/Modules/mpolyquo-localizations.jl [hunk 6]
```diff
@@ -98,3 +98,28 @@ function pre_saturated_module(M::SubQuo{T}) where {T<:MPolyQuoLocalizedRingElem}
...
+    T<:MPolyQuoLocalizedRingElem, 
+    DomType<:FreeMod{T},
+    CodType<:FreeMod{T}
+  }
+  S = base_ring(domain(f))
+  A = representing_matrix(f)
+  B, D = clear_denominators(A)
+  Fb = base_ring_module(domain(f))
+  Gb = base_ring_module(codomain(f))
+  GGb, p = quo(Gb, [g*e for g in gens(modulus(underlying_quotient(S))) for e in gens(Gb)])
```
Comment (afkafkafk13) [inline src/Modules/mpolyquo-localizations.jl:116]: I do not see, where you use GGb later on. You are only aiming for the p here?

Comment (HechtiDerLachs) [reply src/Modules/mpolyquo-localizations.jl:116]: Yes.

**Review by afkafkafk13 — APPROVED** [review]

**Comment by HechtiDerLachs** [comment]
Are we good to merge?

**Pull request merged**